odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2385 @@
+
+        - Note that this definition does **not** fulfill the limit
+          property in :math:`p`, i.e.
+
+          .. math::
+              \| a\|_{c, p} \\not\\to
+              \| a \|_{c, \\infty} \quad (p \\to \\infty)
+
+          unless :math:`c = 1`.
+
+        - The constant must be positive, otherwise it does not define an
+          inner product or norm, respectively.
+        """
+        super().__init__(const, impl='numpy', exponent=exponent,
+                         dist_using_inner=dist_using_inner)
+
+    def inner(self, x1, x2):
+        """Return the weighted inner product of ``x1`` and ``x2``.
+
+        Parameters
+        ----------
+        x1, x2 : `NumpyTensor`
+            Tensors whose inner product is calculated.
+
+        Returns
+        -------
+        inner : float or complex
+            The inner product of the two provided tensors.
+        """
+        if self.exponent != 2.0:
+            raise NotImplementedError('no inner product defined for '
+                                      'exponent != 2 (got {})'
+                                      ''.format(self.exponent))
+        else:
+            inner = self.const * _inner_default(x1, x2)
+            return x1.space.field.element(inner)
+
+    def norm(self, x):
+        """Return the weighted norm of ``x``.
+
+        Parameters
+        ----------
+        x1 : `NumpyTensor`
+            Tensor whose norm is calculated.
+
+        Returns
+        -------
+        norm : float
+            The norm of the tensor.
+        """
+        if self.exponent == 2.0:
+            return sqrt(self.const) * float(_norm_default(x))
+        elif self.exponent == float('inf'):
+            return self.const * float(_pnorm_default(x, self.exponent))
+        else:
+            return (self.const ** (1 / self.exponent) *
+                    float(_pnorm_default(x, self.exponent)))
+
+    def dist(self, x1, x2):
+        """Return the weighted distance between ``x1`` and ``x2``.
+
+        Parameters
+        ----------
+        x1, x2 : `NumpyTensor`
+            Tensors whose mutual distance is calculated.
+
+        Returns
+        -------
+        dist : float
+            The distance between the tensors.
+        """
+        if self.dist_using_inner:
+            dist_squared = (_norm_default(x1) ** 2 + _norm_default(x2) ** 2 -
+                            2 * _inner_default(x1, x2).real)
+            if dist_squared < 0.0:  # Compensate for numerical error
+                dist_squared = 0.0
+            return sqrt(self.const) * float(sqrt(dist_squared))
+        elif self.exponent == 2.0:
+            return sqrt(self.const) * _norm_default(x1 - x2)
+        elif self.exponent == float('inf'):
+            return self.const * float(_pnorm_default(x1 - x2, self.exponent))
+        else:
+            return (self.const ** (1 / self.exponent) *
+                    float(_pnorm_default(x1 - x2, self.exponent)))
+
+
+clas
```

Comment:
Can't really make much sense of the comments on this file from here, they all end up being shown with the wrong code lines.